engine/src/main/java/org/camunda/bpm/engine/impl/cmd/SetJobRetriesCmd.java

```diff
@@ -16,54 +16,101 @@
+import org.camunda.bpm.engine.impl.persistence.entity.JobEntity;
 import org.camunda.bpm.engine.impl.persistence.entity.PropertyChange;
 
-import java.io.Serializable;
-
 
 /**
  * @author Askar Akhmerov
  */
-public class SetJobRetriesCmd extends AbstractSetJobRetriesCmd implements Command<Void>, Serializable {
+public class SetJobRetriesCmd implements Command<Void>, Serializable {
 
   protected static final long serialVersionUID = 1L;
 
+  protected static final String RETRIES = "retries";
+  protected static final String DUE_DATE = "dueDate";
 
   protected final String jobId;
   protected final String jobDefinitionId;
+  protected final List<String> jobIds;
   protected final int retries;
+  protected final Date dueDate;
 
-
-  public SetJobRetriesCmd(String jobId, String jobDefinitionId, int retries) {
-    if ((jobId == null || jobId.isEmpty()) && (jobDefinitionId == null || jobDefinitionId.isEmpty())) {
-      throw new ProcessEngineException("Either job definition id or job id has to be provided as parameter.");
-    }
-
-    if (retries < 0) {
-      throw new ProcessEngineException("The number of job retries must be a non-negative Integer, but '" + retries + "' has been provided.");
-    }
-
+  public SetJobRetriesCmd(String jobId, String jobDefinitionId, int retries, Date dueDate) {
     this.jobId = jobId;
     this.jobDefinitionId = jobDefinitionId;
+    this.jobIds = null;
     this.retries = retries;
+    this.dueDate = dueDate;
+  }
+
+  public SetJobRetriesCmd(List<String> jobIds, int retries, Date dueDate) {
+    this.jobId = null;
+    this.jobDefinitionId = null;
+    this.jobIds = jobIds;
+    this.retries = retries;
+    this.dueDate = dueDate;
   }
 
   public Void execute(CommandContext commandContext) {
     if (jobId != null) {
-      setJobRetriesByJobId(jobId, retries, commandContext);
-    } else {
+      setJobRetriesByJobId(jobId, retries, dueDate, commandContext);
+    } else if(jobDefinitionId != null){
       setJobRetriesByJobDefinitionId(commandContext);
+    } else if(jobIds != null) {
+      for (String id : jobIds) {
+        setJobRetriesByJobId(id, retries, dueDate, commandContext);
+      }
     }
 
     return null;
   }
 
+  protected void setJobRetriesByJobId(String jobId, int retries, Date dueDate, CommandContext commandContext) {
+    JobEntity job = commandContext
+        .getJobManager()
+        .findJobById(jobId);
+    if (job != null) {
+      for (CommandChecker checker : commandContext.getProcessEngineConfiguration().getCommandCheckers()) {
+        checker.checkUpdateRetriesJob(job);
+      }
+
+      if (job.isInInconsistentLockState()) {
+        job.resetLock();
+      }
+
+      List<PropertyChange> propertyChanges = new ArrayList<>();
+
+      int oldRetries = job.getRetries();
+      job.setRetries(retries);
+      propertyChanges.add(new PropertyChange(RETRIES, oldRetries, job.getRetries()));
+
+      if(dueDate != null) {
```

Comment:
As far as I can see, there is an engine config property called `ensureJobDueDateNotNull` that is disabled by default. This is respected when the JobExecutor acquires new jobs. If I set this flag to true and a due date to null it won't be executed.

To be honest, this feels like a product decision. I am fine with either way but I guess if we allow `null` due dates here we should consider the case outlined above.

Reply:
I don't see it as a product decision but as making the feature consistent with existing behavior.

Possibilities:
1. when passing `null` and `ensureJobDueDateNotNull` is `true`, throw an exception; when `false`, set the due date to `null`
2. when passing `null` and `ensureJobDueDateNotNull` is `true`, set the due date to the current date; when `false`, set the due date to `null`

I tend to # 1 since this behavior makes things very transparent to the user. However, I would throw the exception already on the creation level of the batch operation. Not on the execution level to avoid failing jobs that need to be resolved.

Reply:
* I don't particularly appreciate that I need to find out myself which solution you opted for by looking into the implementation since your decision and reasoning are not documented here.
* Could you let me know why you went for solution # 2?
* Is it now possible to set the due date to `null`?

Reply:
I realized, that in the [docs](https://docs.camunda.org/manual/7.18/reference/deployment-descriptors/tags/process-engine/#ensureJobDueDateNotNull), the property `ensureJobDueDateNotNull` is described as "If the Job DueDate property hasn't been explicitly set, the current time is added." so I figured we should handle it the same in this case.
[Here is the relevant code for that.](https://github.com/camunda/camunda-bpm-platform/blob/0f84ae8c95a5e4da6a1c0d37a3995d2a2b3ef7cf/engine/src/main/java/org/camunda/bpm/engine/impl/cmd/SetJobRetriesCmd.java#L69-L71)
In that sense if `ensureJobDueDateNotNull` is set to `false` and `null` is passed, the job duedate should not be set.

Sorry, this was not clear and I should have documented my findings here. I also just realized that the documentation states the default value for `ensureJobDueDateNotNull` as `true`, while [it is actually](https://github.com/camunda/camunda-bpm-platform/blob/0f84ae8c95a5e4da6a1c0d37a3995d2a2b3ef7cf/engine/src/main/java/org/camunda/bpm/engine/ProcessEngineConfiguration.java#L215) `false`.

Reply:
* It's fair to make it consistent with what is documented.
* Why do you conclude that if `ensureJobDueDateNotNull` is set to `false`, a `null` job due date should be ignored?
   * Wouldn't it be more intuitive that it can be set to `null` in this case?

Reply:
I will improve the PR to allow `null` due dates.